### PR TITLE
Font table editor

### LINKED ISSUE #85: Add editor features to Font Table viewer
Depends on:
* https://github.com/OpenDiablo2/OpenDiablo2/issues/909
* #75

### PR DESCRIPTION
**Pull request by gucio321**
fixes #85

### CODE REVIEW

File: hswidget/fonttablewidget/widget.go [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+			}),
+			giu.Button("down##"+p.id+"upItem"+string(r)).Size(downW, downH).OnClick(func() {
+				p.itemDown(r)
+			}),
+		),
+		giu.Line(
+			giu.Button("edit##"+p.id+"editRune"+string(r)).Size(editRuneW, editRuneH).OnClick(func() {
+				state.editRune.startRune = r
+				state.mode = fontTableWidgetEditRune
+			}),
```
Comment (gucio321) [inline hswidget/fonttablewidget/widget.go]: :confused: there should be `ImageButtons` but hsutil.MakeImageButton causes panic:
```golang
panic: Assertion failed!
File: /home/user/go/pkg/mod/github.com/ianling/imgui-go@v1.12.1-0.20210115050955-434395a2da82/imgui/imgui.h, Line 1421

Expression: Size > 0
```
exactly it is caused by imgui.PushID()

File: hswidget/fonttablewidget/state.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+type fontTableWidgetMode int
+
+const (
+	// viewer state
+	fontTableWidgetViewer fontTableWidgetMode = iota
+	// edit rune
+	fontTableWidgetEditRune
+	// opens a dialog and adds a new item on first free index
+	fontTableWidgetAddItem
```
Comment (gravestench) [inline hswidget/fonttablewidget/state.go]: since this is in its own package, you can reduce the noise
```golang
const (
	modeViewer fontTableWidgetMode = iota
	modeEditRune
	modeAddItem
)
```

File: hswidget/fonttablewidget/state.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+// Dispose cleans state
+func (s *widgetState) Dispose() {
+	s.editRune.Dispose()
+	s.addItem.Dispose()
+}
+
+type editRune struct {
+	editedRune int32
+	startRune  rune
+}
```
Comment (gravestench) [inline hswidget/fonttablewidget/state.go]: `editRune` sounds like an action instead of a thing. would this make more sense as `runeEditState` ? 

example:
```golang
widget.runeEdit.Dispose()
```
vs
```golang
widget.runeEditState.Dispose()
```

File: hswidget/fonttablewidget/state.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+func (e *editRune) Dispose() {
+	e.editedRune = rune(0)
+	e.startRune = rune(0)
+}
+
+type addItem struct {
+	newRune editRune
+	width,
+	height int32
+}
```
Comment (gravestench) [inline hswidget/fonttablewidget/state.go]: Same goes for this; it sounds like an action instead of a thing.

File: hswidget/fonttablewidget/widget.go [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+	addW, addH               = 400, 30
+	editRuneW, editRuneH     = 50, 30
+	saveCancelW, saveCancelH = 80, 30
+)
+
+type textures struct {
+	up, down, del *giu.Texture
+}
+
+// widget represents a font table widget
```
Comment (gravestench) [inline hswidget/fonttablewidget/widget.go]: we can remove this comment lol

File: hswidget/fonttablewidget/widget.go [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+	fontTable *d2font.Font
+	id        string
+	textures
+}
+
+// Create creates a new FontTable widget
+func Create(
+	up, down, del *giu.Texture,
+	id string, fontTable *d2font.Font,
+) giu.Widget {
```
Comment (gravestench) [inline hswidget/fonttablewidget/widget.go]: I think in the future we should try to leave these textures out of the provider function for the widget. These are more or less hardcoded, so why not just keep these inside of the widget impl?

If you look at the widgets exported by `giu`, I think you'll see a lot of them will take a uid and some other widget parameters in the provider func; these textures are not really parameters in the sense that they change widget operation, they are literally being passed in to be assigned somewhere.

tldr; we should manage textures inside of the widget, not pass them into the provider func.

**Review by gravestench — CHANGES_REQUESTED** [review]
can you change `editRune` to `stateEditRune` and `addItem` to `stateAddItem`

File: hswidget/fonttablewidget/widget.go (hunk 1 above)
Comment (gravestench) [reply hswidget/fonttablewidget/widget.go]: have you tried running the debugger to see what is causing the panic?

File: hswidget/fonttablewidget/state.go (hunk 2 above)
Comment (gucio321) [reply hswidget/fonttablewidget/state.go]: solved in 18e5eef

File: hswidget/fonttablewidget/state.go (hunk 3 above)
Comment (gucio321) [reply hswidget/fonttablewidget/state.go]: solved by 7d29fa5

File: hswidget/fonttablewidget/state.go (hunk 4 above)
Comment (gucio321) [reply hswidget/fonttablewidget/state.go]: same here: 7d29fa5

File: hswidget/fonttablewidget/widget.go (hunk 5 above)
Comment (gucio321) [reply hswidget/fonttablewidget/widget.go]: 4945d62

File: hswidget/fonttablewidget/widget.go (hunk 1 above)
Comment (gucio321) [reply hswidget/fonttablewidget/widget.go]: I can't see absolutly nothing, what could cause this crash. in cof editor works fine, but here...

Comment (gucio321) [reply hswidget/fonttablewidget/widget.go]: but I'm working on workaround

Comment (gucio321) [reply hswidget/fonttablewidget/widget.go]: a95c37f

**Comment by gucio321** [comment]
@gravestench could you review again please?

File: hscommon/hsutil/imagebuttonhack.go [hunk 7]
```diff
@@ -2,23 +2,19 @@ package hsutil
...
 	"github.com/ianling/giu"
-	"github.com/ianling/imgui-go"
 )
 
-// MakeImageButton is a hack for giu.ImageButton that manually sets the ImageButton ID by
-// manually setting the ID through imgui
-//nolint:unparam // width and height are always 15 at the time of writing, but may change
+// it looks like childWidget's size must be a bit greater then widget iside of it
+const (
+	stdMod = 8
```
Comment (gravestench) [inline hscommon/hsutil/imagebuttonhack.go]: what does `stdMod` stand for...?

Comment (gucio321) [reply hscommon/hsutil/imagebuttonhack.go]: hmm. right, it isn't right name. it comes from `standard modifier`

File: hswidget/fonttablewidget/widget.go (hunk 6 above)
Comment (gucio321) [reply hswidget/fonttablewidget/widget.go]: it doesn't work;
when we place texture loading in widget, image buttons aren't displayed.

File: hscommon/hsutil/imagebuttonhack.go (hunk 7 above)
Comment (gravestench) [reply hscommon/hsutil/imagebuttonhack.go]: i figured, but the variable is local to this file, you can be more descriptive.

File: hswidget/fonttablewidget/widget.go (hunk 6 above)
Comment (gucio321) [reply hswidget/fonttablewidget/widget.go]: maybe, we can place it in widget's state. What do you think about it?

File: hscommon/hsutil/imagebuttonhack.go (hunk 7 above)
Comment (gucio321) [reply hscommon/hsutil/imagebuttonhack.go]: da0a135

File: hscommon/hsutil/imagebuttonhack.go [hunk 8]
```diff
@@ -2,23 +2,19 @@ package hsutil
...
-			// so it wont be unique across multiple instances if we use the same texture...
-			// we need to step over giu and manually tell imgui to pop the last ID and
-			// push the desired one onto the stack
-			imgui.PopID()
-			imgui.PushID(id)
-		}),
+		giu.Child(id+"child").Border(false).Size(float32(w+childSizeModifier), float32(h+childSizeModifier)).Layout(giu.Layout{
+			giu.ImageButton(t).Size(float32(w), float32(h)).OnClick(fn),
+		}).Flags(giu.WindowFlagsNoDecoration),
 	}
```
Comment (gravestench) [inline hscommon/hsutil/imagebuttonhack.go]: just my opinion, but i find what you have here hard to read. There's no reason why we can't just have a few more lines of code and comment what we're doing _so that other people can read it_.
```golang
	const (
		childIdSuffix = "child"
		padding       = 8 // pixels
	)

	// the button
	btnW, btnH := float32(w), float32(h)
	button := giu.Layout{
		giu.ImageButton(t).Size(btnW, btnH).OnClick(fn),
	}

	// the container; needs to be padded to be larger than the button
	id := fmt.Sprintf("%s%s", id, childIdSuffix)
	con := giu.Child(id).
		Border(false).
		Size(float32(btnW+padding), float32(btnH+padding)).
		Layout(button).
		Flags(giu.WindowFlagsNoDecoration)

	return con
```

File: hswidget/fonttablewidget/widget.go [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+	)
+
+	p.fontTable.Glyphs[state.addItemState.newRune] = newGlyph
+	state.mode = modeViewer
+}
+
+// this giant custom function allows us to
+// check if letter entered by user already exists in map
+// end depending on it, build save and cancel buttons
+// or cancel only
```
Comment (gravestench) [inline hswidget/fonttablewidget/widget.go]: stale comment

**Review by gravestench — DISMISSED** [review]

File: hswidget/fonttablewidget/widget.go (hunk 9 above)
Comment (gucio321) [reply hswidget/fonttablewidget/widget.go]: a4aab17

File: hscommon/hsutil/imagebuttonhack.go (hunk 8 above)
Comment (gucio321) [reply hscommon/hsutil/imagebuttonhack.go]: why not: 0c63ad1

File: hswidget/fonttablewidget/widget.go (hunk 6 above)
Comment (gravestench) [reply hswidget/fonttablewidget/widget.go]: Yeah, we might be able to, but it seems almost like an app-level piece of functionality (loading single texture to be used everywhere). Perhaps we can come up with an enum for the texture ID's + a mapping of paths to ID's. Then we might be able to put all the texture loading stuff somewhere else and not worry about that in the widget code.

anyways, definitely another PR.

**Review by gravestench — APPROVED** [review]
